File: ios/flutter_eco_mode/Sources/flutter-eco-mode/listeners/BatteryStateListener.swift
//
//  BatteryStateListener.swift
//  flutter_eco_mode
//
//  Created by CHOUPAULT Alexis on 27/02/2026.
//

import Flutter

class BatteryStateListener: BatteryStateStreamHandler, DisposableStreamListener {
    private var eventSink: PigeonEventSink<BatteryState>?
    private var lastSentState: BatteryState?
    
    // MARK: - BatteryStateStreamHandler implementation

    override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<BatteryState>) {
        self.eventSink = sink
        
        sendUpdate()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(batteryStateChanged),
            name: UIDevice.batteryStateDidChangeNotification,
            object: nil
        )
    }
    
    override func onCancel(withArguments arguments: Any?) {
        cleanUp()
    }
    
    // MARK: - DisposableStreamListener implementation
    
    func register(binaryMessenger: FlutterBinaryMessenger) {
        BatteryStateStreamHandler.register(with: binaryMessenger, streamHandler: self)
    }
    
    func dispose(binaryMessenger: FlutterBinaryMessenger) {
        cleanUp()
        
        FlutterEventChannel(name: EcoModeEventChannels.batteryState, binaryMessenger: binaryMessenger, codec: messagesPigeonMethodCodec).setStreamHandler(nil)
    }
    
    // MARK: - Private methods

    @objc private func batteryStateChanged() {
        sendUpdate()
    }

    private func sendUpdate() {
        let currentState = EcoBatteryManager.shared.getBatteryState()
        
        guard currentState != lastSentState else { return }
        
        lastSentState = currentState
        
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?.success(currentState)
        }
    }
    
    private func cleanUp() {
        NotificationCenter.default.removeObserver(self)
        eventSink = nil
        lastSentState = nil
    }
}

File: android/src/main/kotlin/sncf/connect/tech/flutter_eco_mode/listener/BatteryStateListener.kt
package sncf.connect.tech.flutter_eco_mode.listener

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import sncf.connect.tech.flutter_eco_mode.BatteryState
import sncf.connect.tech.flutter_eco_mode.BatteryStateStreamHandler
import sncf.connect.tech.flutter_eco_mode.EcoBatteryManager
import sncf.connect.tech.flutter_eco_mode.MessagesPigeonMethodCodec
import sncf.connect.tech.flutter_eco_mode.PigeonEventSink

class BatteryStateListener(
    private val ecoBatteryManager: EcoBatteryManager,
) : BatteryStateStreamHandler(), DisposableStreamListener {

    private var batteryStateEventSink: PigeonEventSink<BatteryState>? = null
    private var batteryStateReceiver: BroadcastReceiver? = null
    private var lastSentState: BatteryState? = null

    override fun onListen(p0: Any?, sink: PigeonEventSink<BatteryState>) {
        batteryStateEventSink = sink

        batteryStateReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent?) {
                sendBatteryStatus(intent)
            }
        }

        val stickyIntent = ecoBatteryManager.registerBatteryReceiver(batteryStateReceiver)
        sendBatteryStatus(stickyIntent)
    }

    override fun onCancel(p0: Any?) {
        cleanUp()
    }

    override fun register(binaryMessenger: BinaryMessenger) = register(binaryMessenger, this)

    override fun dispose(binaryMessenger: BinaryMessenger) {
        cleanUp()

        EventChannel(binaryMessenger, EcoModeEventChannels.BATTERY_STATE, MessagesPigeonMethodCodec).setStreamHandler(null)
    }

    private fun cleanUp() {
        batteryStateReceiver?.let { ecoBatteryManager.unregisterReceiver(it) }
        batteryStateReceiver = null
        batteryStateEventSink = null
        lastSentState = null
    }

    private fun sendBatteryStatus(intent: Intent?) {
        val currentState = ecoBatteryManager.parseState(intent)

        if (currentState != lastSentState) {
            lastSentState = currentState
            batteryStateEventSink?.success(currentState)
        }
    }
}
